utils/s2n_random.c

```diff
@@ -128,65 +130,74 @@ S2N_RESULT s2n_get_mix_entropy(struct s2n_blob *blob)
-        /* Instantiate the new ones */
-        RESULT_GUARD(s2n_drbg_instantiate(&per_thread_public_drbg, &public, S2N_AES_128_CTR_NO_DF_PR));
-        RESULT_GUARD(s2n_drbg_instantiate(&per_thread_private_drbg, &private, S2N_AES_128_CTR_NO_DF_PR));
-        zero_if_forked_ptr = zeroed_when_forked_page;
-        zero_if_forked = 1;
-    }
+    RESULT_GUARD(s2n_drbg_instantiate(&s2n_per_thread_rand_state.public_drbg, &public, S2N_AES_128_CTR_NO_DF_PR));
+    RESULT_GUARD(s2n_drbg_instantiate(&s2n_per_thread_rand_state.private_drbg, &private, S2N_AES_256_CTR_NO_DF_PR));
+
+    s2n_per_thread_rand_state.drbgs_initialized = true;
 
     return S2N_RESULT_OK;
 }
 
-S2N_RESULT s2n_get_public_random_data(struct s2n_blob *blob)
+static S2N_RESULT s2n_maybe_init_drbgs(void)
 {
-    RESULT_GUARD(s2n_defend_if_forked());
+    if (s2n_per_thread_rand_state.drbgs_initialized == false) {
+        RESULT_GUARD(s2n_init_drbgs());
 
-    uint32_t offset = 0;
-    uint32_t remaining = blob->size;
+        /* Then cache the fork generation number. We just initialized the drbg
+         * states with new entropy and forking is not an external event.
+         */
+        uint64_t returned_fork_generation_number = 0;
+        RESULT_GUARD(s2n_get_fork_generation_number(&returned_fork_generation_number));
+        s2n_per_thread_rand_state.cached_fork_generation_number = returned_fork_generation_number;
+    }
 
-    while(remaining) {
-        struct s2n_blob slice = { 0 };
+    return S2N_RESULT_OK;
+}
 
-        RESULT_GUARD_POSIX(s2n_blob_slice(blob, &slice, offset, MIN(remaining, S2N_DRBG_GENERATE_LIMIT)));;
+/* s2n_ensure_uniqueness() implements defenses against uniqueness
+ * breaking events that might cause duplicated drbg states. Currently, only
+ * implements fork detection.
+ */
+static S2N_RESULT s2n_ensure_uniqueness(void)
+{
+    uint64_t returned_fork_generation_number = 0;
+    RESULT_GUARD(s2n_get_fork_generation_number(&returned_fork_generation_number));
 
-        RESULT_GUARD(s2n_drbg_generate(&per_thread_public_drbg, &slice));
+    if (returned_fork_generation_number != s2n_per_thread_rand_state.cached_fork_generation_number) {
 
-        remaining -= slice.size;
-        offset += slice.size;
+        /* This assumes that s2n_rand_cleanup_thread() doesn't mutate any other
+         * state than the drbg states and it resets the drbg initialization
+         * boolean to false. s2n_maybe_init_drbgs() will cache the new fork
+         * generation number in the per thread state.
+         */
+        RESULT_GUARD(s2n_rand_cleanup_thread());
+        RESULT_GUARD(s2n_maybe_init_drbgs());
     }
 
     return S2N_RESULT_OK;
 }
 
-S2N_RESULT s2n_get_private_random_data(struct s2n_blob *blob)
+static S2N_RESULT s2n_get_random_data(struct s2n_blob *out_blob,
+    struct s2n_drbg *drbg_state)
 {
-    RESULT_GUARD(s2n_defend_if_forked());
+    RESULT_GUARD(s2n_maybe_init_drbgs());
+    RESULT_GUARD(s2n_ensure_uniqueness());
```

Comment:
ensure_uniqueness() calls s2n_maybe_init_drbgs(). Is this intentional?

Reply:
yes, because `s2n_rand_cleanup_thread()` wipes the drbg.

You could maybe do

```
RESULT_GUARD(s2n_ensure_uniqueness());
RESULT_GUARD(s2n_maybe_init_drbgs());
```
and get rid of the init. But I would rather handle it correctly inside `s2n_ensure_uniqueness()` and not having to rely on it.